docs/root/configuration/http/http_filters/adaptive_concurrency_filter.rst

```diff
@@ -0,0 +1,197 @@
+
+Concurrency Limit Headroom
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+The headroom value is necessary as a driving factor to increase the concurrency limit when the
+sampleRTT is in the same ballpark as the minRTT. This value must be present in the limit
+calculation, since it forces the concurrency limit to increase until there is a deviation from the
+minRTT latency. In the absence of a headroom value, the concurrency limit could potentially stagnate
+at an unnecessary small value if the sampleRTT and minRTT are close to each other.
+
+Because the headroom value is so necessary to the proper function for the gradient controller, the
+headroom value is unconfigurable and pinned to the square-root of the concurrency limit.
+
+Limitations
+-----------
+The adaptive concurrency filter's control loop relies on latency measurements
+and adjustments to the concurrency limit based on those measurements. Because of
+this, the filter must operate in conditions where it has full control over
+request concurrency. This means that:
+
+    1. The filter works as intended in the filter chain for a local cluster.
+
+    2. The filter must be able to limit the concurrency for a cluster. This means
+       there must not be requests destined for a cluster that are not decoded by
+       the adaptive concurrency filter.
+
+Example Configuration
+---------------------
+An example filter configuration can be found below. Not all fields are required and many of the
+fields can be overridden via runtime settings.
+
+.. code-block:: yaml
+
+  name: envoy.filters.http.adaptive_concurrency
+  config:
+    gradient_controller_config:
+      sample_aggregate_percentile:
+        value: 90
+      concurrency_limit_params:
+        concurrency_update_interval: 0.1s
+      min_rtt_calc_params:
+        jitter:
+          value: 10
+        interval: 60s
+        request_count: 50
+    enabled:
+      default_value: true
+      runtime_key: "adaptive_concurrency.enabled"
+
+The above configuration can be understood as follows:
+
+* Gather latency samples for a time window of 100ms. When entering a new window, summarize the
+  requests (sampleRTT) and and update the concurrency limit using this sampleRTT.
+* When calculating the sampleRTT, use the p90 of all sampled latencies for that window.
+* Recalculate the minRTT every 60s and add a jitter (random delay) of 0s-6s to the start of the
+  minRTT recalculation. The delay is dictated by the jitter value.
+* Collect 50 request samples to calculate the minRTT and use the p90 to summarize them.
+* The filter is enabled by default.
+
+.. note::
+
+    It is recommended that the adaptive concurrency filter come after the healthcheck filter in the
+    filter chain to prevent latency sampling of health checks. If health check traffic is sampled,
+    it could potentially affect the accuracy of the minRTT measurements.
+
+Runtime
+-------
+
+The adaptive concurrency filter supports the following runtime settings:
+
+adaptive_concurrency.enabled
```

Comment:
It occurs to me that if we ever want to support both ingress and egress adaptive concurrency in a single side car, we will have conflicting runtime names. Would it be better to not hard code these names and instead read them from runtime value configuration fields or similar? We might consider doing this in a follow up before we consider this filter production ready. WDYT?

Reply:
Since these runtime parameters are all overrides of the config parameters, we can just use the runtime configuration fields for the config like you mention. That'll allow for unique runtime names.

Let's knock that out in a different patch. I'll open an issue.